feat: allow restricting public file uploads to System Managers

manzoorsofi opened the pull request:
This PR introduces a new security setting that helps prevent accidental public file exposure in environments with many non-technical users

🔒 New System Setting

**System Settings** → Files -> _Only allow System Managers to upload public files_

When enabled:

✔ Only System Managers (and Administrator) can upload public files
✔ Other users cannot toggle “Private” off
✔ Other users do not see:

“Is Private” checkbox in File Uploader

“Set all private / Set all public” actions in upload dialog

✔ Backend validation also prevents users from making a file public via API or form

### Upload dialog

as System Manager (setting ON or OFF):

<img width="1210" height="587" alt="Screenshot 2025-11-25 104500" src="https://github.com/user-attachments/assets/594e79c4-59ea-442f-9dfd-97af4305ea3d" />

as normal user (setting ON):

<img width="580" height="246" alt="Bildschirmfoto 2025-12-10 um 01 26 31" src="https://github.com/user-attachments/assets/884fd162-2a32-4019-a0ce-7143b3585864" />

### File DocType 

as normal user (setting ON):

<img width="2457" height="1133" alt="Screenshot 2025-11-25 114129" src="https://github.com/user-attachments/assets/11afef3f-cffa-4d03-9433-1168efc40c01" />

(The _Is Private_ field is read-only if already private)

### Backward Compatibility
✔ Default = OFF → no change for existing installations
✔ Fully backward compatible with all permissions


### Documentation
Docs: https://docs.frappe.io/framework/

manzoorsofi committed:
feat: restrict public file uploads to System Managers

manzoorsofi committed:
chore: trigger ci

manzoorsofi committed:
chore: trigger ci

manzoorsofi commented:
Vulnerable Dependency Check is failing due to pypdf==6.1.3 (GHSA-m449-cwjh-6pw7). This dependency version comes from the base branch; this PR doesn’t modify any Python dependencies

Please let me know if you’d like me to handle the pypdf bump in this PR or if that will be addressed separately.

barredterra commented on frappe/core/doctype/file/file.js:
```diff
@@ -37,6 +37,16 @@ frappe.ui.form.on("File", {
+		const roles = frappe.user_roles || [];
+		const isSystemManager = roles.includes("System Manager");
+
+		if (onlySysMgr && !isSystemManager) {
```
```suggestion
		if (
			frappe.boot.sysdefaults?.only_system_managers_upload_public_files &&
			!frappe.user.has_role("System Manager")
		) {
```

barredterra replied:
Also apply to frappe/public/js/frappe/file_uploader/file_uploader.bundle.js

barredterra commented on frappe/core/doctype/file/file.py:
```diff
@@ -156,6 +157,14 @@ def validate_attachment_references(self):
+		):
+			roles = frappe.get_roles()
+			if "System Manager" not in roles:
+				frappe.throw(_("Public files can only be created by System Managers."))
```
```suggestion
			frappe.only_for("System Manager")
```

barredterra reviewed (commented):
In general, we can refactor the check for sysdefaults and role into a utility function `can_upload_public_files`.

barredterra commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -461,6 +463,13 @@
+    "fieldtype": "Check",
+    "label": "Only allow System Managers to upload public files",
+    "description": "If enabled, only System Managers can upload public files. Other users can't see the checkbox <i>Is Private</i> in the upload dialog."
+  }, 
```
In general, please only modify DocType definitions through the UI. Do not edit the JSON files directly. This ensures that the JSON files are accurate.

barredterra replied:
@manzoorsofi DocType definitions need to be adjusted via UI. This was clearly not done here. Please fix.

manzoorsofi committed:
feat: restrict public file uploads to System Managers

manzoorsofi committed:
refactor: add can_upload_public_files util

barredterra commented on frappe/core/doctype/file/file.js:
```diff
@@ -37,6 +37,12 @@ frappe.ui.form.on("File", {
+		if (!frappe.file_utils.can_upload_public_files() && frm.doc.is_private) {
+			// Only make read-only if the file is currently private
+			frm.set_df_property("is_private", "read_only", 1);
+		}
```
```suggestion
		if (!frappe.file_utils.can_upload_public_files() && frm.doc.is_private) {
			frm.set_df_property("is_private", "read_only", 1);
		}
```

These comments are redundant

barredterra commented on frappe/core/doctype/file/file.py:
```diff
@@ -156,6 +157,10 @@ def validate_attachment_references(self):
 
+	def enforce_public_file_restrictions(self):
+		if not cint(self.is_private) and not can_upload_public_files():
+			frappe.only_for("System Manager")
```
```suggestion
			try:
				frappe.only_for("System Manager")
			except PermissionError:
				frappe.throw(_("Only System Managers can make this file public."))
```

Actually this seems better. Sorry for the incomplete earlier suggestion.

barredterra commented on frappe/core/doctype/file/file.py:
```diff
@@ -156,6 +157,10 @@ def validate_attachment_references(self):
 			frappe.throw(_("The fieldname you've specified in Attached To Field is invalid"))
 
+	def enforce_public_file_restrictions(self):
+		if not cint(self.is_private) and not can_upload_public_files():
```
```suggestion
		if not self.is_private and frappe.get_system_settings("only_system_managers_upload_public_files"):
```

I would think that both are always integers in this context. Did you observe something else?

`can_upload_public_files` is redundant here, it duplicates the logic of `frappe.only_for`.

barredterra commented on frappe/core/doctype/file/utils.py:
```diff
@@ -440,3 +440,10 @@ def find_file_by_url(path: str, name: str | None = None) -> Optional["File"]:
+	"""Return True if current user can upload public files."""
+	if not cint(frappe.get_system_settings("only_system_managers_upload_public_files")):
+		return True
+	return "System Manager" in frappe.get_roles()
```
```suggestion
```

not needed

barredterra commented on frappe/public/js/frappe/file_uploader/utils.js:
I'm not sure if this is a good location for this function. Having it in the `file_uploader` folder implies that it's specific to `FileUploader`, but it's not.

manzoorsofi committed:
refactor: move can_upload_public_files to frappe.utils and address review feedback

manzoorsofi committed:
chore: merge upstream/develop and resolve conflicts

manzoorsofi commented:
Merge Conflicts:The disable_product_suggestion (system settings) field has been successfully merged from develop.

manzoorsofi committed:
test: add tests for public file upload restriction

barredterra commented on cypress/integration/file_uploader.js:
```diff
@@ -52,4 +52,104 @@ context("FileUploader", () => {
+				doctype: "System Settings",
+				field: "only_allow_system_managers_to_upload_public_files",
+				value: 1,
+			});
```
This seems to be hallucinated. `cy.set_value` should be the correct approach.

manzoorsofi committed:
test: fix Cypress tests to use cy.set_value for system settings

manzoorsofi committed:
test: fix and refactor Cypress tests for system settings

manzoorsofi committed:
Merge branch 'develop' into feat/restrict-public-file-upload

manzoorsofi committed:
fix: merge conflict resolution for System Settings

barredterra reviewed (approved):
## Review Summary

(AI-generated review) Thank you for this security feature! It addresses a real need in environments with non-technical users.

### CI Status
All checks passing.

### What works well
- Server-side validation using `frappe.only_for("System Manager")` - correct Frappe pattern
- Validation runs in `validate()` which covers both insert and update operations
- Comprehensive test coverage (Python unit tests + Cypress E2E tests)
- Clean client-side utility function `can_upload_public_files()` centralized in utils.js
- Good UX: hiding/disabling controls prevents futile attempts that would fail server-side anyway
- Default OFF maintains backward compatibility

### Minor suggestion (non-blocking)

Would it be worth adding an explicit test for the update scenario - verifying that an existing private file cannot be converted to public by a non-System Manager when the setting is enabled? The current tests cover file creation well, but an update test would document that behavior explicitly.

Looks good!

barredterra committed:
Merge remote-tracking branch 'upstream/develop' into feat/restrict-public-file-upload

barredterra committed:
fix: separate display and perms

One property is for hiding the checkbox when state is uploaded or failed. The other is for whether the user should be able to change the value. If they shouldn't, the checkbox now stays visible but read-only.

barredterra committed:
fix: uploaded files cannot be public by default when that is not allowed

For some doctypes, uploaded files are marked as public by default. When the user lacks perms to upload public files, this needs to be prevented.

barredterra commented:
Tested, LGTM.

Failing test cases don't seem to be our fault. Merge `develop` in some time, after they're fixed.

barredterra committed:
Merge branch 'develop' into feat/restrict-public-file-upload

barredterra committed:
fix: add IDs to "optimize" and "private" checkbox labels in FilePreview

This makes them easier to identify in test cases.

barredterra committed:
test: check for enabled/disabled

Before, we checked for visible vs missing.
